server/src/test/java/org/opensearch/cluster/service/ClusterManagerTaskThrottlerTests.java

```diff
@@ -406,6 +412,143 @@ public void testThrottling() {
         throttler.onBeginSubmit(getMockUpdateTaskList(taskKey, throttlingKey, 1));
     }
 
+    public void testThrottlingWithLock() {
+        ClusterManagerThrottlingStats throttlingStats = new ClusterManagerThrottlingStats();
+        String taskKey = "test";
+        ClusterSettings clusterSettings = new ClusterSettings(Settings.builder().build(), ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+        ClusterManagerTaskThrottler throttler = new ClusterManagerTaskThrottler(Settings.EMPTY, clusterSettings, () -> {
+            return clusterService.getClusterManagerService().getMinNodeVersion();
+        }, throttlingStats);
+        ClusterManagerTaskThrottler.ThrottlingKey throttlingKey = throttler.registerClusterManagerTask(taskKey, true);
+
+        throttler.updateLimit(taskKey, 5);
+
+        // adding 3 tasks
+        throttler.onBeginSubmit(getMockUpdateTaskList(taskKey, throttlingKey, 3));
+
+        // adding 3 more tasks, these tasks should be throttled
+        // taskCount in Queue: 3 Threshold: 5
+        assertThrows(
+            ClusterManagerThrottlingException.class,
+            () -> throttler.onBeginSubmit(getMockUpdateTaskList(taskKey, throttlingKey, 3))
+        );
+        assertEquals(3L, throttlingStats.getThrottlingCount(taskKey));
+
+        // remove one task
+        throttler.onBeginProcessing(getMockUpdateTaskList(taskKey, throttlingKey, 1));
+
+        // add 3 tasks should pass now.
+        // taskCount in Queue: 2 Threshold: 5
+        throttler.onBeginSubmit(getMockUpdateTaskList(taskKey, throttlingKey, 3));
+
+        final CountDownLatch latch = new CountDownLatch(1);
+        // Taking lock on tasksCount will not impact throttling behaviour now.
+        var threadToLock = new Thread(() -> {
+            throttler.tasksCount.computeIfPresent(taskKey, (key, count) -> {
+                try {
+                    latch.await();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                return 10L;
+            });
+        });
+        threadToLock.start();
+
+        // adding one task will throttle
+        // taskCount in Queue: 5 Threshold: 5
+        final ClusterManagerThrottlingException exception = assertThrows(
+            ClusterManagerThrottlingException.class,
+            () -> throttler.onBeginSubmit(getMockUpdateTaskList(taskKey, throttlingKey, 1))
+        );
+        assertEquals("Throttling Exception : Limit exceeded for test", exception.getMessage());
+        assertEquals(Optional.of(5L).get(), throttler.tasksCount.get(taskKey));
+        assertEquals(4L, throttlingStats.getThrottlingCount(taskKey));
+
+        latch.countDown();
+        try {
+            // Wait to complete and then assert on new tasksCount that got modified by threadToLock Thread
+            threadToLock.join();
```

Comment:
In case above assertions fail, this thread will be left orphan. Same for below test as well. you will need to clean them up

Reply:
Made changes to clean up threads.